orttraining/orttraining/python/training/ortmodule/_logger.py

```diff
@@ -76,3 +77,95 @@ class LogColor:
+    CREATE_SESSION = 5  # The latency of creating the session
+
+    def to_string(self) -> str:
+        if self == TimeTrackerPhase.EndToEnd:
+            return "end to end"
+        if self == TimeTrackerPhase.EXPORT:
+            return "export"
+        elif self == TimeTrackerPhase.GRAPH_BUILDER_INIT:
+            return "graph builder init"
+        elif self == TimeTrackerPhase.DETECTION:
+            return "runtime detection"
+        elif self == TimeTrackerPhase.BUILD_GRAPH:
+            return "graph building"
+        elif self == TimeTrackerPhase.CREATE_SESSION:
+            return "session creation"
+        else:
+            return "invalid"
+
+
+class TimeTracker:
+    """A simple class to track time spent in different phases of ORT backend first-time initialization."""
+
+    NOT_RECORD = -1.0
+
+    def __init__(
+        self,
+    ):
+        self.starts_: List[float] = [TimeTracker.NOT_RECORD] * len(TimeTrackerPhase)
+        self.ends_: List[float] = [TimeTracker.NOT_RECORD] * len(TimeTrackerPhase)
+
+    def start(self, phase: TimeTrackerPhase):
+        self.starts_[phase] = time.time()
+
+    def end(self, phase: TimeTrackerPhase):
+        self.ends_[phase] = time.time()
+
+    def _get_duration(self, phase: TimeTrackerPhase):
+        if self.ends_[phase] == TimeTracker.NOT_RECORD or self.starts_[phase] == TimeTracker.NOT_RECORD:
+            return TimeTracker.NOT_RECORD
+        return self.ends_[phase] - self.starts_[phase]
+
+    def to_string(self, log_details=False) -> str:
+        end_to_end_str = self._get_duration(TimeTrackerPhase.EndToEnd)
+        end_to_end_str = f"{end_to_end_str:.2f}" if end_to_end_str != TimeTracker.NOT_RECORD else "N/A"
+        export_str = self._get_duration(TimeTrackerPhase.EXPORT)
+        export_str = f"{export_str:.2f}" if export_str != TimeTracker.NOT_RECORD else "N/A"
+        overhead_title_str = (
+            f"Total ORT initialization overhead is {end_to_end_str}s where export takes {export_str}s.\n"
+        )
+
+        if log_details is False:
+            return overhead_title_str
+
+        duration_summaries = []
+        for phase in TimeTrackerPhase:
+            _get_duration = self._get_duration(phase)
+            if phase in [TimeTrackerPhase.EndToEnd, TimeTrackerPhase.EXPORT]:
+                continue
+
+            val = (
+                f" {phase.to_string()} takes {_get_duration:.2f}s" if _get_duration != TimeTracker.NOT_RECORD else "N/A"
+            )
+            duration_summaries.append(f"{val}")
+
+        return f"{overhead_title_str}Other overhead details: {','.join(duration_summaries)}\n"
+
+
+class TrackTime:
+    """A function decorator to track time spent in different phases of ORT backend first-time initialization."""
+
+    def __init__(self, phase: TimeTrackerPhase):
+        self.phase = phase
+
+    def __call__(self, func: Callable):
+        def wrapper(*args, **kwargs):
```

Comment:
```suggestion
        def wrapper(graph_execution_manager: GraphExecutionManager, *args, **kwargs):
```

And then use that everywhere as opposed to `_self`?

Reply:
OK, I thought about it then it will limit that only usable for training_manager and inference_manager, but should not be a problem. Let me fix is as you suggested. thanks man!

Reply:
did not add the type hint for GraphExecutionManager because this will cause a recursive import between training_manager and _logger.py.